File: src/main/java/com/example/Nadeuri/member/security/util/JWTUtil.java
package com.example.Nadeuri.member.security.util;

import com.example.Nadeuri.member.dto.JwtTokenDTO;
import io.jsonwebtoken.*;

import io.jsonwebtoken.security.Keys;
import lombok.Data;
import lombok.extern.log4j.Log4j2;
import org.apache.catalina.User;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.io.UnsupportedEncodingException;
import java.time.Duration;
import java.util.Arrays;
import java.util.Collection;
import java.util.Date;
import java.util.Map;
import java.util.stream.Collectors;

@Component
@Log4j2
public class JWTUtil {
    private static String key//서명에 사용할 키 - 30자 이상으로
            = "1234567890123456789012345678901234567890123";

    // Member 정보를 가지고 AccessToken, RefreshToken을 생성하는 메서드
    public JwtTokenDTO createToken(Authentication authentication ) {
        SecretKey key = null;
        try {
            key = Keys.hmacShaKeyFor(JWTUtil.key.getBytes("UTF-8"));
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }

        // 인증된 사용자의 권한 목록을 쉼표로 구분된 문자열로 변환(권한 가져오기)
        String authorities = authentication.getAuthorities().stream()
                .map(GrantedAuthority::getAuthority)
                .collect(Collectors.joining(","));

        long now = (new Date()).getTime();


        //Access 토큰 생성
        Date accessTokenExpiresIn = new Date(now + 86400000);
        String accessToken = Jwts.builder()
                .setSubject(authentication.getName())
                .claim("auth", authorities)
                .setExpiration(accessTokenExpiresIn)
                .signWith(key)
                .compact();

        //Refresh 토큰 생성
        String refreshToken = Jwts.builder()
                .setExpiration(new Date(now + 86400000))
                .signWith(key, SignatureAlgorithm.HS256)
                .compact();

        return JwtTokenDTO.builder()
                .grantType("Bearer")
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .build();
    }

    //Jwt 토큰을 복호화하여 토큰에 들어있는 정보를 꺼내는 메서드
    public Authentication getAuthentication(String Token) {
        //Jwt 토큰 복호화
        Claims claims = parseClaims(Token);

        if (claims.get("auth") == null) {
            throw new RuntimeException("권한 정보가 없는 토큰입니다.");
        }

        // 클레임에서 권한 정보 가져오기
        Collection<? extends GrantedAuthority> authorities = Arrays.stream(claims.get("auth").toString().split(","))
                .map(SimpleGrantedAuthority::new)
                .collect(Collectors.toList());

        // UserDetails 객체를 만들어서 Authentication return
        // UserDetails: interface, User: UserDetails를 구현한 class
        UserDetails principal = new User(claims.getSubject(), "", authorities);
        return new UsernamePasswordAuthenticationToken(principal, "", authorities);
    }



    //검증 기능 수행
    public boolean validateToken(String token) {
        SecretKey key = null;
        try {
            key = Keys.hmacShaKeyFor(JWTUtil.key.getBytes("UTF-8"));
        } catch (UnsupportedEncodingException e) {
            throw new RuntimeException(e);
        }

        try {
            Jwts.parser()
                    .setSigningKey(key)
                    .build()
                    .parseClaimsJws(token);
            return true;
        } catch (SecurityException | MalformedJwtException e) {
            log.info("Invalid JWT Token", e);
        } catch (ExpiredJwtException e) {
            log.info("Expired JWT Token", e);
        } catch (UnsupportedJwtException e) {
            log.info("Unsupported JWT Token", e);
        } catch (IllegalArgumentException e) {
            log.info("JWT claims string is empty.", e);
        }
        return false;
    }


    // 주어진 액세스 토큰을 파싱하여 클레임을 반환하고, 만료된 경우 클레임을 반환하는 메서드
    private Claims parseClaims(String accessToken) {
        try {
            return Jwts.parser()
                    .setSigningKey(key)
                    .build()
                    .parseClaimsJws(accessToken)
                    .getBody();
        } catch (ExpiredJwtException e) {
            return e.getClaims();
        }
    }
}

File: src/main/java/com/example/Nadeuri/member/security/util/JWTUtil.kt
package com.example.Nadeuri.member.security.util

import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import lombok.extern.log4j.Log4j2
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.io.UnsupportedEncodingException
import java.nio.charset.StandardCharsets
import java.time.Duration
import java.util.*
import javax.crypto.SecretKey

@Component
@Log4j2
class JWTUtil {

    @Value("\${jwt.secret.key}")
    private lateinit var key: String // 서명에 사용할 키 - 30자 이상으로

    fun createToken(valueMap: Map<String, Any>, min: Int): String {
        val secretKey = Keys.hmacShaKeyFor(key.toByteArray(StandardCharsets.UTF_8))

        val now = Date() // 토큰 발행 시간
        return Jwts.builder()
            .setHeaderParam("alg", "HS256") // HS256 알고리즘으로 헤더에 alg 필드로 추가
            .setHeaderParam("type", "JWT") // 헤더에 타입 필드를 추가하고 값으로 jwt 설정
            .setIssuedAt(now) // 토큰 발행 시간
            .setExpiration(Date(now.time + Duration.ofMinutes(min.toLong()).toMillis())) // 토큰 만료 시간
            .setClaims(valueMap) // 저장 데이터 (jwt 페이로드에 저장되는 사용자 정보와 같은 추가 데이터)
            .signWith(secretKey) // 서명(서명을 위한 비밀키를 key로 설정)
            .compact() // 헤더, 페이로드, 서명이 결합되어 jwt 문자열을 만듬
    }

    // 검증 기능 수행
    fun validateToken(token: String): Map<String, Any> {
        val secretKey = Keys.hmacShaKeyFor(key.toByteArray(StandardCharsets.UTF_8))

        val claims: Claims = Jwts.parser()
            .setSigningKey(secretKey)
            .build()
            .parseClaimsJws(token)
            .body

        return claims
    }
}